Add relative time option to simulated sensors

Shou opened the pull request:
## Description:

By default simulated sensors are relative to when they're activated,
instead we make this togglable with this new option 'relative_to_epoch',
and instead they relative to 1970-01-01 00:00:00.

**Related issue (if applicable):**

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#5223

## Example entry for `configuration.yaml` (if applicable):
```yaml
sensor:
  - platform: simulated
    name: 'one'
    unit: '%'
    amplitude: 256
    mean: 0
    spread: 0
    seed: 0
    period: 86400
    relative_to_epoch: true
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [ ] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ ] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

balloob commented on homeassistant/components/sensor/simulated.py:
```diff
@@ -84,11 +91,14 @@ def __init__(self, name, unit, amp, mean, period, phase, fwhm, seed):
         """"Return the time delta."""
         dt0 = self._start_time
+        if self._absolute_time:
+            dt0 = datetime.datetime(1970, 1, 1, tzinfo=pytz.utc)
```
Please use `homeassistant.util.dt.UTC` as tzinfo

Shou replied:
I've updated that now.

balloob commented on homeassistant/components/sensor/simulated.py:
```diff
@@ -49,6 +52,7 @@
     vol.Optional(CONF_PHASE, default=DEFAULT_PHASE): vol.Coerce(float),
     vol.Optional(CONF_FWHM, default=DEFAULT_FWHM): vol.Coerce(float),
     vol.Optional(CONF_SEED, default=DEFAULT_SEED): cv.positive_int,
+    vol.Optional(CONF_ABSOLUTE_TIME, default=DEFAULT_ABSOLUTE_TIME): cv.boolean
```
I don't think that absolute time is the correct description. It's a description of the value, not what the value is used for.

Shou replied:
What do you propose?

balloob commented:
CC @robmarkcole

robmarkcole commented:
@Shou another suggested name for the variable could be `relative_to_epoch`

Shou commented:
Thanks! I've renamed the variables now.

balloob commented:
relative to epoch still has the same problem as absolute time. Quoting myself from my comment above:

>  It's a description of the value, not what the value is used for.

Shou commented:
Hey @robmarkcole @balloob, sorry for the late reply – my intent here is to be able to use simulated sensors to create smooth curves of values that I can feed into, e.g. lightbulbs and have them turn on and off slowly depending on the time of day, so indeed I am trying to synchronise with a daily event. Maybe `day_synchronised` or similar is a better variable name for intent then?

robmarkcole commented:
@Shou can that not be achieved using a regular automation including the `sin` you added in [this PR](https://github.com/home-assistant/home-assistant/pull/13915)?

Shou commented:
@robmarkcole absolutely, but then with this I get the benefits of the sensor too.

robmarkcole commented:
Well its OK by me

balloob commented:
Actually, thinking here, why wouldn't we ALWAYS make it relative to epoch?

robmarkcole commented:
Agreed, this could default to true, which means you would  have a reliable periodic signal for dimming/brightening lights througout the day. The user would just have to determine the required phase offset for their timezone.

Shou commented:
Yeah, you could essentially get the behaviour of 'relative to start time' with an offset, albeit it would get a bit annoying to change all the time should you want the feature.

So to verify, I should make this PR enable this by default? Do we still want to change the variable name?

balloob commented:
I would go one step further, why make this configurable at all ? I think that this PR should just set start time to UNIX epoch and call it a day. No config options.

Shou commented:
While this sounds good to me, @robmarkcole is there a case where we wouldn't want to remove 'relative to now'?

robmarkcole commented:
Well my original reason for developing this component was to enable testing of filters and automations. Therefore I want reproducible behaviour each time I restart HA

balloob commented:
@robmarkcole I don't think I understand. Relative to epoch makes it continue at the same value as before the restart, making it reproducible?

balloob commented:
Can anyone elaborate why you would want it to restart from 0 the moment you restart Home Assistant instead of continue at where it left off?

robmarkcole commented:
Relative to epoch means it starts at a point that is dependent on the time. 

My particular use for having a reproducible start point, would be if I were doing a demo of an automation that was triggered by a rising signal that passes a threshold

balloob commented:
> Relative to epoch means it starts at a point that is dependent on the time.

Isn't it the exact opposite? By making it relative to epoch, it always starts at exactly the same time instead of when Home Assistant was restarted, making it independent of time?

balloob commented:
I'm fine with it either way. Just making sure we really need an extra config option.

Shou committed:
add relative time option to simulated sensors

By default simulated sensors are relative to when they're activated,
instead we make this togglable with this new option 'relative_to_epoch',
and instead they become relative to 1970-01-01 00:00:00.

balloob commented on homeassistant/components/sensor/simulated.py:
```diff
@@ -79,11 +87,14 @@ def __init__(self, name, unit, amp, mean, period, phase, fwhm, seed):
     def time_delta(self):
         """Return the time delta."""
         dt0 = self._start_time
+        if self._relative_to_epoch:
```
This is the wrong implementation. We only need to calculate `self._start_time` once, and that's in the constructor.

Shou replied:
Thanks, I should've fixed this now.

Shou commented:
Sorry for the noise here, I've now updated this to be `relative_to_epoch` by default, tested it locally, and updated the attached documentation PR.

balloob commented:
🎉 🎉 🎉 🎉 🎉 🎉 

awesome! Thanks!
